LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -88,46 +85,39 @@ private bool EnsureFirstMessageInCache(string key, LoRaRequest loRaRequest, out
             return false;
         }
 
-        internal static string CreateCacheKey(LoRaPayloadData payload)
+        internal static string CreateCacheKey(LoRaPayloadData payload, LoRaDevice loRaDevice)
         {
-            var totalBufferLength = DevAddr.Size + Mic.Size + (payload.RawMessage?.Length ?? 0) + payload.Fcnt.Length;
-            var buffer = totalBufferLength <= 128 ? stackalloc byte[totalBufferLength] : new byte[totalBufferLength]; // uses the stack for small allocations, otherwise the heap
-
-            var index = 0;
-            _ = payload.DevAddr.Write(buffer);
-            index += DevAddr.Size;
-
-            _ = payload.Mic is { } someMic ? someMic.Write(buffer[index..]) : throw new InvalidOperationException("Mic must not be null.");
-            index += Mic.Size;
-
-            payload.RawMessage?.CopyTo(buffer[index..]);
-            index += payload.RawMessage?.Length ?? 0;
-
-            BinaryPrimitives.WriteUInt16LittleEndian(buffer[index..], BinaryPrimitives.ReadUInt16LittleEndian(payload.Fcnt.Span));
+            var totalBufferLength = DevEui.Size + Mic.Size + payload.Fcnt.Length;
+            Span<byte> buffer = stackalloc byte[totalBufferLength];
+            var head = buffer; // keeps a view pointing at the start of the buffer
 
-            var key = Sha256.ComputeHash(buffer.ToArray());
+            buffer = DevEui.Parse(loRaDevice.DevEUI.AsSpan()).Write(buffer);
```

Comment:
Conversion to Span<char> is not needed as it's implicit:

```suggestion
            buffer = DevEui.Parse(loRaDevice.DevEUI).Write(buffer);
```